lib/src/signing.rs

```
@@ -0,0 +1,217 @@
+pub enum SignError {
+    #[error("Would override a foreign commit signature")]
+    CannotRewrite,
+    #[error("Signing error: {0}")]
+    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum SignBehavior {
+    /// Drop existing signatures.
+    /// This is what jj did before signing support or does now when a signing
+    /// backend is not configured.
+    #[default]
+    Drop,
+    /// Only sign commits that were authored by self and already signed,
+    /// "preserving" the signature across rewrites.
+    /// This is what jj does when a signing backend is configured.
+    Keep,
+    /// Sign/re-sign commits that were authored by self. Foreign signatures are
+    /// overwritten on authored commits and dropped on others.
+    /// This is what jj does when configured to always sign.
+    Own,
+    /// Always sign commits, regardless of who authored or signed them before.
+    /// This is what jj does on `jj sign -f` or `jj --force-sign <snapshotting
+    /// command>`.
+    Force,
+}
+
+impl SignBehavior {
+    pub fn from_config(sign_all: bool) -> Self {
+        match sign_all {
+            true => SignBehavior::Own,
+            false => SignBehavior::Keep,
+        }
+    }
+}
+
+/// Wraps low-level signing backends and adds caching, similar to `Store`.
+#[derive(Debug, Default)]
+pub struct Signer {
+    main_backend: Option<Box<dyn SigningBackend>>,
+    other_backends: Vec<Box<dyn SigningBackend>>,
+    cache: RwLock<HashMap<CommitId, Verification>>,
+}
+
+impl Signer {
+    pub fn from_settings(settings: &UserSettings) -> Self {
+        let mut other_backends: Vec<Box<dyn SigningBackend>> = vec![
+            // todo implement some backends
+        ];
+
+        let main_backend = settings
+            .signing_backend()
+            .and_then(|backend| other_backends.iter().position(|b| b.name() == backend))
+            .map(|pos| other_backends.swap_remove(pos));
```

nit: might be easier to build a map of `name: backend`. We'll probably need to report an error if the configured backend name is wrong.

(Error handling can be added later.)

---

Meh, with there being like 3 backends total (maybe a bit more but still a tiny number) a linear search is perfectly fine if not better.

Although with a map there's no need to have `SigningBackend::name` anymore, hmm

For now I made it into a map, but btw I feel like some out-of-git cache of verifications will have to be made at some point in the future as an optimisation to not call the backend (aka subprocess into `gpg` of `ssh-keygen` or whatever) for every single signed commit - right now if some repo had a large number of those the performance would be awful (I'm planning to add the signature checkmark to default commit log template)